src/precomputed_quantities.jl

```diff
@@ -319,18 +370,238 @@ function set_precomputed_quantities!(Y, p, t)
         end
     end
 
+    if turbconv_model isa DiagnosticEDMFX
+        FT = eltype(Y)
+        (; ᶜp_ref, ᶜρ_ref, ᶠu³) = p
+        (; q_tot) = p.ᶜspecific
+        (; ᶜρaʲs, ᶠu³ʲs, ᶜuʲs, ᶜKʲs, ᶜh_totʲs, ᶜq_totʲs, ᶜtsʲs, ᶜρʲs) = p
+        ᶜ∇p_perturb³ = p.ᶜtemp_CT3
+        ᶜ∇Φ³ = p.ᶜtemp_CT3_2
+
+        @. ᶜ∇p_perturb³ = CT3(ᶜgradᵥ(ᶠinterp(ᶜp - ᶜp_ref)))
+        @. ᶜ∇p_perturb³ += CT3(gradₕ(ᶜp - ᶜp_ref))
+        @. ᶜ∇Φ³ = CT3(ᶜgradᵥ(ᶠinterp(ᶜΦ)))
+        @. ᶜ∇Φ³ += CT3(gradₕ(ᶜΦ))
+
+        ρaʲu³ʲ_data = p.temp_data_level
+        ρaʲu³ʲ_datau³ʲ_data = p.temp_data_level_2
+        ρaʲu³ʲ_datah_tot = ρaʲu³ʲ_dataq_tot = p.temp_data_level_3
+
+        for j in 1:n
+            ᶜρaʲ = ᶜρaʲs.:($j)
+            ᶠu³ʲ = ᶠu³ʲs.:($j)
+            ᶜKʲ = ᶜKʲs.:($j)
+            ᶜh_totʲ = ᶜh_totʲs.:($j)
+            ᶜtsʲ = ᶜtsʲs.:($j)
+            ᶜρʲ = ᶜρʲs.:($j)
+            ᶜq_totʲ = ᶜq_totʲs.:($j)
+
+            # boundary condition
+            ρaʲ_level = Fields.field_values(Fields.level(ᶜρaʲ, 1))
+            u³ʲ_halflevel = Fields.field_values(Fields.level(ᶠu³ʲ, half))
+            Kʲ_level = Fields.field_values(Fields.level(ᶜKʲ, 1))
+            h_totʲ_level = Fields.field_values(Fields.level(ᶜh_totʲ, 1))
+            q_totʲ_level = Fields.field_values(Fields.level(ᶜq_totʲ, 1))
+            tsʲ_level = Fields.field_values(Fields.level(ᶜtsʲ, 1))
+            ρʲ_level = Fields.field_values(Fields.level(ᶜρʲ, 1))
+            ρ_level = Fields.field_values(Fields.level(Y.c.ρ, 1))
+            uₕ_level = Fields.field_values(Fields.level(Y.c.uₕ, 1))
+            h_tot_level = Fields.field_values(Fields.level(ᶜh_tot, 1))
+            q_tot_level = Fields.field_values(Fields.level(q_tot, 1))
+
+            p_level = Fields.field_values(Fields.level(ᶜp, 1))
+            Φ_level = Fields.field_values(Fields.level(ᶜΦ, 1))
+
+            local_geometry_level = Fields.field_values(
+                Fields.level(Fields.local_geometry_field(Y.c), 1),
+            )
+            local_geometry_halflevel = Fields.field_values(
+                Fields.level(Fields.local_geometry_field(Y.f), half),
+            )
+
+            @. u³ʲ_halflevel = CT3(FT(0))
+            @. ρaʲ_level = ρ_level * FT(0.1)
```

Comment:
Would it be at all useful to have `0.1` as parameter and not hard coded?

Reply:
I think so. I put it [here](https://github.com/CliMA/ClimaAtmos.jl/blob/a6ed6d10ea53beb91ae56482ca3790d0a60b12e0/src/solver/types.jl#L106) for now. I think the idea is we want to set this to zero when we have entrainment? But actually, I'm confused about how we can get updraft when looking at the equations. If we need this number in the end, we can add it to ClimaParameters.